File: tools/lib/station-taxonomy.py
"""Approved station categories shared with the JavaScript importer via JSON policy."""
import html
import json
import re
import unicodedata
from pathlib import Path
ROOT = Path(__file__).resolve().parents[2]

def canonical(value):
    return re.sub(r'\s+', ' ', unicodedata.normalize('NFKC', html.unescape(value)).lower()).strip()

def classify(tags, country='gr'):
    policy = load_policy(country)
    genres, formats, aliases, remove = policy
    result = {'genres': [], 'formats': [], 'review': [], 'removed': []}
    def add(key, value):
        if value not in result[key]: result[key].append(value)
    def visit(raw, seen=frozenset()):
        if not isinstance(raw, str): return
        value = canonical(raw)
        if not value or value in remove or re.fullmatch(r'\d+\s*(kbps|kbit/s|bit)', value):
            if value: add('removed', value)
            return
        if re.fullmatch(r'[0-9]0|(?:19|20)[0-9]0', value): value += 's'
        match = re.fullmatch(r"(?:19|20)?(\d0)[’'´]?s", value)
        if match and not value.startswith(('192','193','194')): value = match[1] + 's'
        if value in genres: add('genres', value); return
        if value in formats: add('formats', value); return
        if value in aliases and value not in seen:
            for target in aliases[value]: visit(target, seen | {value})
            return
        parts = re.split(r'[,;|]', value)
        if len(parts)>1:
            for part in parts: visit(part, seen)
            return
        add('review', value)
    for tag in tags or []: visit(tag)
    return result

from functools import lru_cache
@lru_cache(maxsize=None)
def load_policy(country):
    policy=json.loads((ROOT/'tools/config/station-taxonomy.json').read_text())
    shared=json.loads((ROOT/'tools/config/genre-cleanup.json').read_text())
    path=ROOT/f'countries/{country}.genre-cleanup.json'
    custom=json.loads(path.read_text()) if path.exists() else {}
    aliases={canonical(k): [canonical(v) for v in (values if isinstance(values,list) else [values])]
             for k,values in {**shared['aliases'],**policy['aliases'],**custom.get('aliases',{})}.items()}
    remove={canonical(v) for v in shared['remove']+policy['remove']+custom.get('remove',[])}
    return set(policy['genres']),set(policy['formats']),aliases,remove

File: tools/lib/station-taxonomy.mjs
import { readFileSync, existsSync } from 'node:fs';
const root = new URL('../../', import.meta.url);
const read = path => JSON.parse(readFileSync(new URL(path, root), 'utf8'));
const policy = read('tools/config/station-taxonomy.json');
const shared = read('tools/config/genre-cleanup.json');
const cache = new Map();
const entities = {amp:'&',quot:'"',apos:"'",lt:'<',gt:'>',nbsp:' '};
export const canonical = value => value.replace(/&(#x[0-9a-f]+|#\d+|amp|quot|apos|lt|gt|nbsp);/gi, (raw,key) => {
  if (key.startsWith('#')) {
    const n=key[1].toLowerCase()==='x'?parseInt(key.slice(2),16):Number(key.slice(1));
    return n>0 && n<=0x10ffff ? String.fromCodePoint(n) : raw;
  }
  return entities[key.toLowerCase()] || raw;
}).normalize('NFKC').toLowerCase().replace(/\s+/gu,' ').trim();
export function classifyTags(tags, country='gr') {
  if (!cache.has(country)) {
    const url=new URL(`countries/${country}.genre-cleanup.json`,root);
    const custom=existsSync(url)?JSON.parse(readFileSync(url,'utf8')):{};
    cache.set(country, {
      aliases:Object.fromEntries(Object.entries({...shared.aliases,...policy.aliases,...custom.aliases}).map(([k,v])=>[canonical(k),(Array.isArray(v)?v:[v]).map(canonical)])),
      remove:new Set([...shared.remove,...policy.remove,...(custom.remove||[])].map(canonical)),
    });
  }
  const {aliases,remove}=cache.get(country);
  const result={genres:[],formats:[],review:[],removed:[]};
  const add=(key,value)=>{if(!result[key].includes(value))result[key].push(value);};
  function visit(raw,seen=new Set()) {
    if(typeof raw!=='string')return;
    let value=canonical(raw);
    if(!value || remove.has(value) || /^\d+\s*(kbps|kbit\/s|bit)$/.test(value)){if(value)add('removed',value);return;}
    if(/^(?:[0-9]0|(?:19|20)[0-9]0)$/.test(value))value+='s';
    const decade=value.match(/^(?:19|20)?(\d0)[’'´]?s$/);
    if(decade && !/^(192|193|194)/.test(value))value=decade[1]+'s';
    if(policy.genres.includes(value)){add('genres',value);return;}
    if(policy.formats.includes(value)){add('formats',value);return;}
    if(Object.hasOwn(aliases,value) && !seen.has(value)){
      for(const target of aliases[value])visit(target,new Set([...seen,value]));return;
    }
    const parts=value.split(/[,;|]/);
    if(parts.length>1){for(const part of parts)visit(part,seen);return;}
    add('review',value);
  }
  for(const tag of tags||[])visit(tag);
  return result;
}
